homeassistant/components/yamaha/media_player.py

```diff
@@ -79,78 +77,74 @@
-    else:
-        ctrl_url = f"http://{host}:80/YamahaRemoteControl/ctrl"
-        receivers = rxv.RXV(ctrl_url, name).zone_controllers()
-
-    devices = []
-    for receiver in receivers:
-        if receiver.zone in zone_ignore:
-            continue
-
-        device = YamahaDevice(name, receiver, source_ignore, source_names, zone_names)
-
-        # Only add device if it's not already added
-        if device.zone_id not in hass.data[DATA_YAMAHA]:
-            hass.data[DATA_YAMAHA][device.zone_id] = device
-            devices.append(device)
+    known_zones = hass.data.setdefault(DATA_YAMAHA, set())
+
+    def _discovery():
+        """Discover receivers from configuration or network."""
+
+        name = config.get(CONF_NAME)
+        host = config.get(CONF_HOST)
+        source_ignore = config.get(CONF_SOURCE_IGNORE)
+        source_names = config.get(CONF_SOURCE_NAMES)
+        zone_ignore = config.get(CONF_ZONE_IGNORE)
+        zone_names = config.get(CONF_ZONE_NAMES)
+
+        if discovery_info is not None:
+            name = discovery_info.get("name")
+            model = discovery_info.get("model_name")
+            ctrl_url = discovery_info.get("control_url")
+            desc_url = discovery_info.get("description_url")
+            receivers = rxv.RXV(
+                ctrl_url, model_name=model, friendly_name=name, unit_desc_url=desc_url
+            ).zone_controllers()
+            _LOGGER.debug("Receivers: %s", receivers)
+            # when we are dynamically discovered config is empty
+            zone_ignore = []
+        elif host is None:
+            receivers = []
+            for recv in rxv.find():
+                receivers.extend(recv.zone_controllers())
         else:
-            _LOGGER.debug("Ignoring duplicate receiver: %s", name)
-
-    def service_handler(service):
-        """Handle for services."""
-        entity_ids = service.data.get(ATTR_ENTITY_ID)
-
-        devices = [
-            device
-            for device in hass.data[DATA_YAMAHA].values()
-            if not entity_ids or device.entity_id in entity_ids
-        ]
+            ctrl_url = f"http://{host}:80/YamahaRemoteControl/ctrl"
+            receivers = rxv.RXV(ctrl_url, name).zone_controllers()
+
+        entities = []
+        for receiver in receivers:
+            if receiver.zone in zone_ignore:
+                continue
+
+            entity = YamahaDevice(
+                name, receiver, source_ignore, source_names, zone_names
+            )
+
+            # Only add device if it's not already added
+            if entity.zone_id not in known_zones:
+                known_zones.add(entity.zone_id)
+                entities.append(entity)
+            else:
+                _LOGGER.debug("Ignoring duplicate receiver: %s", name)
 
-        for device in devices:
-            port = service.data[ATTR_PORT]
-            enabled = service.data[ATTR_ENABLED]
+        return entities
```

Comment:
Please move the entity creation out of the discovery function and return `receivers` here instead.